docs/adr/adr-013-blocksync-overhaul-part-2.md

```diff
@@ -0,0 +1,418 @@
+storage subsystem of the new block synchronization protocol of Celestia's Data Availability network, while the Part II
+focuses on the networking.
+
+The new design drastically speeds up EDS retrieval in __happy path__ by streaming ODS as a single file and recomputing
+parity data forming EDS. Similar approach is taken for ND retrieval. The main goal is to make EDS and ND retrieval
+consistently outpace EDS production for the max EDS size for historical and recent EDSes.
+
+The existing EDS retrieval logic(TODO: Link something) remains a __fallback path__, if the __happy path__ fails due to lack of connected
+peers or the coordinated data withholding attack. With the __fallback path__ retrieval leverages all connected peers,
+including LNs, and granulary requests EDS pieces - __shares__, increasing the likelihood of EDS reconstruction, while
+paying a slower retrieval time price.
+
+## Decision
+
+The decision is to design three protocols:
+- [`ShrEx/EDS`](#ShrEx/EDS Protocol) - Pull-based EDS retrieval protocol
+- 
+- [`ShrEx/ND`](#ShrEx/ND Protocol) - Pull-based ND retrieval protocol
+- 
+- [`ShrEx/Sub`](#ShrEx/Sub Protocol) - Push-based _notification_ protocol for the newly produced EDSes
+
+### Key Design Decisions
+
+- __Use [go-libp2p][go-libp2p]'s software stack to build the protocols.__ It's extensively used in `celestia-node` already,
+proven to be robust and provides all the requirRed components and tools for the new protocols.
+
+- __Use [protobuf][proto] schema for protocol(TODO Link) message (de)serialization.__ Protobuf is a common (de)serialization
+standard which is already used in celestia-node. Even thought there are better alternatives, their benefits are
+negligible comparing with continuing Protobuf usage for consistency and simplicity.
+
+- __FN/LNs keep existing EDS retrieval logic(TODO Link?).__ `celestia-node` already provides means to get EDS from
+the network. It is slower than block production for the EDS sizes __>128__, but has higher retrieval guarantees during
+data withholding attack. (link)
+
+- __FN/LN/BNs leverage implemented discovery mechanism outlined in ADR-008(TODO Link).__ The discovery in a node maintains connections
+with FN-subnetwork. This ensures the node will receive new notifications from `ShrEx/Sub` and can request data via
+`ShrEx/EDS` and `ShrEx/ND`.
+
+- __FN/BNs notify connected LN/FNs when they are ready to serve a recent EDS using via [`ShrEx/Sub`](#ShrEx/Sub Protocol).__
+This lazy-push approach provides reliable EDS dissemination properties and turns out to be simpler than technics similar
+to Bitswap's wantlist.(TODO link)
+
+## Protocols Specs
+
+### ShrEx/EDS Protocol
+
+ShrEx/EDS is a pull-based protocol with client-server model, where LNs are clients, BNs are servers
+and FNs are both. The protocol has only one request-response interaction followed by EDS stream.
+
+The protocol is designed over plain libp2p streams with `/shrex/eds/0.0.1` as the protocol ID.
```

Comment:
> Is it because EDS data will use a different encoder from proto?

Yes. we could put the bytes field into the response, but that would require us to read the whole EDS in mem.

Reply:
> Or there is valid scenario when client takes OK response, but terminates data stream?

Interesting case. This is a scenario of a bad client, I'll add the `Misbehavior` section to the protocol and elaborate on it.

Reply:
We might want to do something about this case if the data stream is terminated repeatedly. If it happens one time only, that might be any kind of externality, like a laptop that went out of battery.

Reply:
This misbehavior is more like a DDoS case, where a pretty valid client just spams us with the correct requests and then cuts the stream, not sure if we want to solve this issue in this particular case or we may come up with more generalized solution with the whole libp2p community

Reply:
Thank you for answer. I agree that sending large chunk of data by single proto message is not a good idea. 

I think possible attack vector could be mostly mitigated by setting proper server side write deadlines.